docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,444 @@
+	- Derive a key from device key and salt "AUTH" + `unique_key_id` field
+	- Allocate and initialize `auth_handle` CMAC calculation handle field with derived key 
+	- Using `auth_handle` handle, calculate CMAC on `key` & `metadata`
+- Call `_underlying_kv` `set_start` API 	    
+- Call `_underlying_kv` `set_add_data` API with `metadata` field 	    
+- Return OK   
+
+**set_add_data function**
+
+Header:  
+`virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);`
+
+Pseudo code:
+
+- If `offset` + `data_size` > data size in handle, return error
+- If flags include encryption
+	- Iterate over `value_data` field in chunks of `_scratch_buf` size
+		- Using `enc_handle` handle field, encrypt chunk into `_scratch_buf`
+		- If authentication flag set, using `auth_handle` handle field, update CMAC of `_scratch_buf`
+		- Call `_underlying_kv` `set_add_data` API with `_scratch_buf` 	    
+- Else
+	- If authentication flag set, using `auth_handle` handle field, update CMAC of `value_data`
+	- Call `_underlying_kv` `set_add_data` API with `value_data`
+- Update `offset` field in handle 	    
+- Return OK   
+
+**set_finalize function**
+
+Header:  
+`virtual int set_finalize(set_handle_t handle);`
+
+Pseudo code:
+
+- Initialize a local `cmac` 16-byte array to 0.
+- If authentication flag set, using `auth_handle` handle field, generate `cmac`
+- Call `_underlying_kv` `set_add_data` API with `cmac`
+- Call `_underlying_kv` `set_finalize`
+- If rollback protect flag set, Call `_underlying_kv` with `_unique_key_id_counter` as key and `cmac` as data
+- Deinitialize and free `auth_handle` and `enc_handle`
+- Free `handle`
+- Release `_mutex`
+- Return OK   
+
+### Key iterator APIs
+
+**iterator_open function**
+
+Header:  
+`virtual int iterator_open(iterator_t *it, const char *prefix = NULL);`
+
+Pseudo code:
+- Allocate a `key_iterator_handle_t` structure into `it`
+- Take `_mutex`  
+- Call `_underlying_kv` `iterator_open` with `underlying_it` field
+- Release `_mutex`
+- Return OK   
+
+**iterator_next function**
+
+Header:  
+`virtual int iterator_next(iterator_t it, char *key, size_t key_size, size_t *actual_key_size);`
+
+Pseudo code:
+- Take `_mutex`  
+- Call `_underlying_kv` `iterator_next` with `underlying_it` field
+- Release `_mutex`
+- Return OK   
+
+**iterator_close function**
+
+Header:  
+`virtual int iterator_close(iterator_t it);`
+
+Pseudo code:
+- Take `_mutex`  
+- Call `_underlying_kv` `iterator_close` with `underlying_it` field
+- Release `_mutex`
+- Deallocate `it`
+- Return OK   
+
+
+# Usage scenarios and examples
+
+### Standard usage of the class
+
+Following example code shows standard usage of the SecureStore class 
+
+**Standard usage example**
+
+```C++
+// Underlying key value store - here TDBStore (should be instantiated and initialized)
+extern TDBStore tdbstore;
+
+// Instantiate SecureStore with tdbstore as underlying key value store
+SecureStore secure_store(&tdbstore);
```

Comment:
What about rbp_tdbstore for rollback protection? it is optional but how do you pass it?

Reply:
You are right, will add.